### PR TITLE
Respect `@Option` annotations on interfaces

### LINKED ISSUE #18496: @Option on interfaces (implemented by tasks) is not respected
Custom tasks can implement custom interfaces (or abstract classes). Gradle respects all task input and output annotations (such as `@Input`, `@InputFile`, etc), _except_ for `@Option`. @big-guy told me that this was probably an oversight. Currently `@Option` is only respected if it's in the task itself, or an abstract super _class_ (not interface).

### Expected Behavior
Gradle should respect `@Option` inputs that are defined in an interface to a task.

### Current Behavior
Gradle does not do so.

### Context

When we hit this bug, we thought that `@Option`s were just special and ended up duplicating a variety of options across many tasks. It didn't occur to us to try to change the interfaces involved to abstract classes instead. Since this is (apparently) the only task input/output that doesn't work in interfaces, it is painful to have to remember this weird caveat.

### Steps to Reproduce 

Given this setup:

```kotlin
interface MyInterface {
  @get:Option(
  option = "serial",
  description = "Target the device with given serial (ANDROID_SERIAL)"
  )
  @get:Optional
  @get:Input
  val serial: Property<String?>
}

abstract class MyTask : DefaultTask(), MyInterface {
  @Action fun action() {
    println("Hello world")
  }
}

class MyPlugin {
  override fun apply(target: Project) {
    target.tasks.register("myTask", MyTask::class.java)
  }
}
```
Execute the task, passing an option:
```
./gradlew myTask --serial 1234
```

Gradle will fail, saying there is no option named "serial".

Change the interface to an abstract class, and it should work. But this should not be necessary.

### Your Environment

Build scan URL: n/a

### LINKED ISSUE #18498: Gradle dependency verification breaks Android Studio Gradle sync in an obscure way
Gradle dependency verification breaks Android Studio Gradle sync in an obscure way

### Expected Behavior
It syncs

### Current Behavior
```
Could not resolve all dependencies for configuration ':placeholder-tests:debugAndroidTestRuntimeClasspath'.
Unexpected state for resolution: Unknown
```

### Context
Our project his this issue when trying to enable Gradle dependency verification, and this is the simplified repro case.

### Steps to Reproduce 
https://github.com/liutikas/gradle-bug-crash-android-studio-sync

1. Download project
2. Open it in Android Studio
3. Gradle Sync

In my case I used
```
Android Studio Bumblebee | 2021.1.1 Canary 13
Build #AI-211.7628.21.2111.7762732, built on September 23, 2021
Runtime version: 11.0.11+0-b60-7590822 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.10.46-4rodete1-amd64
GC: G1 Young Generation, G1 Old Generation
Memory: 10240M
Cores: 72
Registry: external.system.auto.import.disabled=true
Non-Bundled Plugins: org.jetbrains.kotlin (211-1.5.31-release-551-AS7442.40), org.intellij.plugins.markdown (211.7142.37)
Current Desktop: Undefined
```

### Your Environment
Build scan URL:
https://scans.gradle.com/s/z6akqavqci2sw

### PR DESCRIPTION
**Pull request by patrickpichler**
With this change, any `Option` annotation on an interface method will be
picked up and made available to be passed in as arguments.

Fixes #18496

### Context
#18496 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/options/OptionsReaderIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          @TaskAction
+          void action() {
+            println "Serial: ${serial.getOrElse('-')}"
+          }
+        }
+
+        tasks.register("myTask", MyTask.class)
+        '''.stripIndent()
+
+        succeeds('myTask', '--serial=1234')
```
Comment (patrickpichler) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/options/OptionsReaderIntegrationTest.groovy:47]: This is a hack, if anybody knows a better way, please let me know.

File: subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy [hunk 2]
```diff
@@ -229,6 +230,40 @@ class OptionReaderTest extends Specification {
         e.message == "No description set on option 'field' at for class 'org.gradle.api.internal.tasks.options.OptionReaderTest\$TestClass8'."
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/18496")
+    def "handles abstract classes with interfaces"() {
```
[the remark this answers is not in the document]
Comment (patrickpichler) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/options/OptionReaderTest.groovy:234]: I've added another integration test to verify that the second provided argument will 
take precedence and another unit test that both options are detected.

**Review by octylFractal — APPROVED** [review]

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by wolfs** [comment]
@bot-gradle test RFM

**Comment by patrickpichler** [comment]
Mhm the tests fail with `Caused by: java.io.FileNotFoundException: https://downloads.gradle-dn.com/distributions-snapshots/gradle-7.3-20210724022245+0000-bin.zip`. 

It doesn't look like this is caused by my change (at least I would not know how). 
Is there anything I can do to, fix the failing tests?

**Comment by octylFractal** [comment]
@bot-gradle test RFM

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Pull request merged**